### PR TITLE
Fix erroneous write misses

### PR DESCRIPTION
**Pull request by lundgren87**
This PR adds the following:

- A fix for the scenario when two or more read misses occur simultaneously causing all but the first read miss to **wrongfully register as write misses** (which also triggers unnecessary local and _remote_ pyxis downgrades and empty page writebacks). The cause of this error was that we detect write misses as any miss on a page that is already present and valid in the cache, which also applies for simultaneous read misses that wait for the cachemutex while another thread fetches the page in question. By checking the origin of the miss (read or write), we can at least filter out these erroneous misses for x86-64 hardware for now.
- A commit fixing a missing `atomic` include detected by GCC11.
- Definition of `NDEBUG` for release (non-debug) builds, as it may be important to test that the assumption of page fault error bits and the `REG_ERR` register holds true for future builds without impacting performance for release builds. _Please dispute this commit if you do not feel it is a good solution._

### CODE REVIEW

**Comment by lundgren87** [comment]
I will queue this for testing on Jenkins at some point later today.

**Review by kostis — DISMISSED** [review]

**Comment by lundgren87** [comment]
> Did the Jenkings runs pass?
> 
> @davidklaftenegger any comments on this before I merge it?

It's running on Jenkins now, I did not get time to set it up before the weekend. Will update with results when it's completed later today.

File: src/backend/mpi/swdsm.cpp [hunk 1]
```diff
@@ -236,15 +236,28 @@ inline std::size_t align_backwards(std::size_t offset, std::size_t size) {
...
+#ifdef REG_ERR
+	/* On x86, get and decode the error number from the context */
+	const ucontext_t* ctx = static_cast<ucontext_t*>(context);
+	auto err_num = ctx->uc_mcontext.gregs[REG_ERR];
+	assert(err_num & X86_PF_USER);	//Assert signal from user space
+	assert(err_num < X86_PF_RSVD); 	//Assert signal is from read or write access
+	/* This could be further decoded by using X86_PF_PROT to detect
+	 * whether the fault originated from no page found (0) or from
+	 * a protection fault (1), but is not needed for this purpose. */
+	bool write_miss = err_num & X86_PF_WRITE; //True for write, false for read
```
Comment (davidklaftenegger) [inline src/backend/mpi/swdsm.cpp:258]: would the later code be easier if we set this variable outside the `#ifdef`, but replicate the "old" behaviour (first=read, second=write) in its value?

File: src/synchronization/global_tas_lock.hpp [hunk 2]
```diff
@@ -11,6 +11,7 @@
 #include "../data_distribution/global_ptr.hpp"
 #include <chrono>
 #include <thread>
+#include <atomic>
```
Comment (davidklaftenegger) [inline src/synchronization/global_tas_lock.hpp:14]: our coding style doesn't say anything about it, but should these includes be in alphabetic order?

File: src/backend/mpi/swdsm.cpp [hunk 3]
```diff
@@ -236,15 +236,28 @@ inline std::size_t align_backwards(std::size_t offset, std::size_t size) {
 	return (offset / size) * size;
 }
 
-void handler(int sig, siginfo_t *si, void *unused){
+void handler(int sig, siginfo_t *si, void *context){
 	UNUSED_PARAM(sig);
-	UNUSED_PARAM(unused);
```
Comment (davidklaftenegger) [inline src/backend/mpi/swdsm.cpp:241]: does the compiler not complain about the unused context when REG_ERR is not defined?

**Review by davidklaftenegger — CHANGES_REQUESTED** [review]
The algorithmic change looks very good.

I am worried about the debug build failing if REG_ERR is undefined, can you test this and make sure the build doesn't break in this scenario?

All requested changes could be dismissed by a comment why they are not needed.

File: src/backend/mpi/swdsm.cpp (hunk 1 above)
Comment (lundgren87) [reply src/backend/mpi/swdsm.cpp]: I have re-written the code slightly according to this suggestion. Write misses are now classified as `read`, `write` or `undefined` with the latter as default, and we make optimizations in the case where `read` or `write` miss knowledge can be used without checking for the definition of `REG_ERR`.

File: src/backend/mpi/swdsm.cpp (hunk 3 above)
Comment (lundgren87) [reply src/backend/mpi/swdsm.cpp:241]: I believe this should now be fixed.

**Comment by lundgren87** [comment]
Updated with slight re-write in accordance with (I hope) David's thoughts. Tests passed on Jason.

**Comment by lundgren87** [comment]
Simplified the logic to avoid false write misses per David's request, currently running on Jenkins.

**Comment by lundgren87** [comment]
Jenkins passed and if no one has any further comments everything should be ready.

**Review by davidklaftenegger — APPROVED** [review]
looks good.

**Pull request merged**